src/gt4py/next/program_processors/runners/dace_iterator/workflow.py

```diff
@@ -164,17 +199,41 @@ class Meta:
 
 
 def convert_args(
-    inp: stages.CompiledProgram,
+    inp: CompiledDaceProgram,
     device: core_defs.DeviceType = core_defs.DeviceType.CPU,
     use_field_canonical_representation: bool = False,
 ) -> stages.CompiledProgram:
-    sdfg_program = cast(CompiledSDFG, inp)
-    on_gpu = True if device == core_defs.DeviceType.CUDA else False
+    sdfg_program = inp.sdfg_program
     sdfg = sdfg_program.sdfg
+    on_gpu = True if device == core_defs.DeviceType.CUDA else False
 
     def decorated_program(
         *args, offset_provider: dict[str, common.Connectivity | common.Dimension]
     ):
+        if sdfg_program._lastargs:
+            # The scalar arguments should be replaced with the actual value; for field arguments,
+            # the data pointer should remain the same otherwise fast-call cannot be used and
+            # the args list needs to be reconstructed.
+            use_fast_call = True
+            scalar_args_table: dict[str, ctypes._SimpleCData] = {}
+            for pos, (param, arg) in enumerate(zip(sdfg.arg_names, args), start=1):
+                if isinstance(arg, common.Field):
+                    storage_type = sdfg.arrays[param].storage
+                    assert isinstance(sdfg_program._lastargs[0][pos], ctypes.c_void_p)
+                    if sdfg_program._lastargs[0][pos].value != get_array_interface_ptr(
+                        arg.ndarray, storage_type
+                    ):
+                        use_fast_call = False
+                        break
+                else:
+                    scalar_args_table[param] = arg
```

Comment:
Are you sure that this works?
Are you sure that this is always a correct ctypes type and not some plain Python `int`?

Reply:
I see, I should do something like in dace code:
```
                elif not isinstance(arg, (ctypes._SimpleCData, ctypes._Pointer)):
                    newargs[i] = actype(arg)
                else:
                    newargs[i] = arg
```